demos/vlasov_poisson_1d/vp1d.py.rst

```diff
@@ -0,0 +1,377 @@
+     'ksp_rtol': 1.0e-8,
+     }
+  nullspace = VectorSpaceBasis(constant=True)
+  phi_solver = LinearVariationalSolver(phi_problem,
+                                       nullspace=nullspace,
+				       solver_parameters=params)
+
+Now we move onto the solver to compute :math:`\partial f/\partial t`. We
+define a symbolic :math:`\Delta t` which we will update later. ::
+  
+  dtc = Constant(0)
+
+The solver will take in ``fstar`` and return :math:`\Delta t\partial f/\partial t` in ``df_out``. ::
+
+  df_out = Function(V)
+
+Now we express the equation in UFL, starting with the left hand side
+bilinear form ::
+  
+  q = TestFunction(V)
+  u = as_vector([v, -phi.dx(0)])
+  n = FacetNormal(mesh)
+  un = 0.5*(dot(u, n) + abs(dot(u, n)))
+  df = TrialFunction(V)
+  df_a = q*df*dx
+
+The problem is defined on an extruded mesh, so the interior facets are
+separated into horizontal and vertical ones. ::
+
+  dS = dS_h + dS_v
+
+Now we build the right hand side linear form. A conditional operator
+is used to deal with the inflow and outflow parts of the exterior
+boundary. Due to the periodic boundary conditions in :math:`x_1`, the only exterior boundary is at the top and bottom of the domain, with measure `ds_tb`. ::
+  
+  df_L = dtc*(div(u*q)*fstar*dx
+     - (q('+') - q('-'))*(un('+')*fstar('+') - un('-')*fstar('-'))*dS
+     - conditional(dot(u, n) > 0, q*dot(u, n)*fstar, 0.)*ds_tb
+      )
+
+We then use this to build a solver. ::
+
+  df_problem = LinearVariationalProblem(df_a, df_L, df_out)
+  df_solver = LinearVariationalSolver(df_problem)
+
+We are getting close to the time loop. We set up some timestepping
+parameters. ::
+  
+  T = 50.0 # maximum timestep
+  t = 0. # model time
+  ndump = 100 # frequency of file dumps
+  dumpn = 0 # dump counter
+  nsteps = 5000
+  dt = T/nsteps
+  dtc.assign(dt)
+
+We set up some :class:`~.Function`\s to store Runge-Kutta stage variables. ::
+  
+  f1 = Function(V)
+  f2 = Function(V)
+
+We set up a ``VTKFile`` object to write output every ``ndump``
+timesteps. ::
+
+  outfile = VTKFile("vlasov.pvd")
+
+We want to output the initial condition, so need to solve for the electrostatic
+potential that corresponds to the initial density. ::
+
+  fstar.assign(fn)
+  phi_solver.solve()
+  outfile.write(fn, phi)
+  phi.assign(.0)
+
+Now we start the timeloop using a lovely progress bar. ::
+
+  for step in ProgressBar("Timestep").iter(range(nsteps)):
+
+Each Runge-Kutta stage involves solving for :math:`\phi` before solving
+for :math:`\partial f/\partial t`. Here is the first stage. ::
+
+  #
+      fstar.assign(fn)
+      phi_solver.solve()
+      df_solver.solve()
+      f1.assign(fn + df_out)
+
+The second stage. ::
+
+  #
+      fstar.assign(f1)
+      phi_solver.solve()
+      df_solver.solve()
+      f2.assign(3*fn/4 + (f1 + df_out)/4)
+
+The third stage. ::
+
+  #
+      fstar.assign(f2)
+      phi_solver.solve()
+      df_solver.solve()
+      fn.assign(fn/3 + 2*(f2 + df_out)/3)
+      t += dt
```

Comment:
Can this be written as a loop with the coefficient array defined beforehand? I don't think we want to be encouraging people to handcode unrolled Runge-Kutta loops.

Reply:
It's a bit fiddly to do that because of the triviality of the first stage, so it will all look a bit cryptic if I do that, and not very pedagogical. This is also how we did it in the original DG advection demo. I think it is harmless.

Reply:
Not that I can't use IRKsome because it isn't an IRK.